internal/licenses/manager.go

```diff
@@ -161,12 +216,232 @@ func (m *manager) GetLicenseByID(ctx context.Context, id string) (License, error
 
 	license, err := m.store.GetLicenseByID(ctx, id)
 	if err != nil {
-		msg := fmt.Errorf("license with ID %s not found", id)
-		slog.Error("failed to fetch license by ID", "licenseID", id, "error", msg)
+		if errors.Is(err, sql.ErrNoRows) {
+			slog.Debug("license not found", "licenseID", id)
+			return License{}, fmt.Errorf("license with ID %s: %w", id, ErrLicenseNotFound)
+		}
 
-		return License{}, msg
+		slog.Debug("failed to fetch license by ID", "licenseID", id, "error", err)
+		return License{}, err
 	}
 
-	slog.Info("fetched license", "licenseID", id)
+	slog.Debug("fetched license successfully", "licenseID", id)
 	return license, nil
 }
+
+func (m *manager) ClaimLicense(ctx context.Context, fingerprint string) (*LicenseOperationResult, error) {
+	tx, err := m.store.BeginTx(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	storeWithTx := m.store.WithTx(tx)
+	defer tx.Rollback()
+
+	node, err := m.fetchOrCreateNode(ctx, storeWithTx, fingerprint)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch or create node: %w", err)
+	}
+
+	claimedLicense, err := storeWithTx.GetLicenseByNodeID(ctx, &node.ID)
+
+	if err == nil {
+		if !m.config.ExtendOnHeartbeat { // if heartbeat is disabled, we can't extend the claimed license
+			slog.Warn("license claim conflict due to heartbeat disabled", "nodeID", node.ID)
+			return &LicenseOperationResult{Status: OperationStatusConflict}, nil
+		}
+
+		if err := storeWithTx.UpdateNodeHeartbeat(ctx, fingerprint); err != nil {
+			return nil, fmt.Errorf("failed to update node heartbeat: %w", err)
+		}
+
+		if err := tx.Commit(); err != nil {
+			return nil, fmt.Errorf("failed to commit transaction: %w", err)
+		}
+
+		if m.config.EnabledAudit {
+			if err := m.store.InsertAuditLog(ctx, "claimed", "license", claimedLicense.ID); err != nil {
+				slog.Warn("failed to insert audit log", "licenseID", claimedLicense.ID, "error", err)
+			}
+		}
+
+		slog.Info("license extended successfully", "licenseID", claimedLicense.ID)
+		return &LicenseOperationResult{
+			License: &claimedLicense,
+			Status:  OperationStatusExtended,
+		}, nil
+	}
+
+	// claim a new license based on the strategy
+	newLicense, err := m.selectLicenseClaimStrategy(ctx, storeWithTx, &node.ID)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			slog.Warn("no licenses available for claim", "nodeID", node.ID)
```

Comment:
Could we add the fingerprint here as well for debugging purposes? The node ID isn't too helpful, since it may no longer exist.

Reply:
Done
Also, I fixed other places